src/chrome/internal/sources/identifiedLoadedSource.ts

```diff
@@ -0,0 +1,78 @@
+import * as fs from 'fs';
+import { IScript } from '../scripts/script';
+import { IResourceIdentifier } from './resourceIdentifier';
+import { ILoadedSource, IScriptMapper, ICurrentScriptRelationshipsProvider as IScriptMapperProvider, ContentsLocation, SourceScriptRelationship, ImplementsLoadedSource, ScriptAndSourceMapper } from './loadedSource';
+import { ILoadedSourceToScriptRelationship } from './loadedSourceToScriptRelationship';
+import _ = require('lodash');
+import { printArray } from '../../collections/printing';
+import { LocationInScript, LocationInLoadedSource } from '../locations/location';
+
+/**
+ * Loaded Source classification:
+ * Is the script content available on a single place, or two places? (e.g.: You can find similar scripts in multiple different paths)
+ *  1. Single: Is the single place on storage, or is this a dynamic script?
+ *      Single path on storage: RuntimeScriptRunFromStorage
+ *      Single path not on storage: DynamicRuntimeScript
+ *  2. Two: We assume one path is from the webserver, and the other path is in the workspace: RuntimeScriptWithSourceOnWorkspace
+ */
+export class IdentifiedLoadedSource<TString extends string = string> implements ILoadedSource<TString> {
+    public [ImplementsLoadedSource]: 'ILoadedSource' = 'ILoadedSource';
+
+    private constructor(
+        public readonly identifier: IResourceIdentifier<TString>,
+        public readonly sourceScriptRelationship: SourceScriptRelationship,
+        private readonly _scriptMapperProvider: IScriptMapperProvider,
+        public readonly contentsLocation: ContentsLocation) { }
+
+    public get url(): TString {
+        return this.identifier.textRepresentation;
+    }
+
+    public scriptMapper(): IScriptMapper {
+        return this._scriptMapperProvider.scriptMapper(this);
+    }
+
+    public isMappedSource(): boolean {
+        return false;
+    }
+
+    public doesScriptHasUrl(): boolean {
+        return true;
+    }
+
+    public isEquivalentTo(source: ILoadedSource<TString>): boolean {
+        return this === source;
+    }
+
+    public toString(): string {
+        return `src:${this.identifier}`;
+    }
+
+    public static create<TString extends string>(identifier: IResourceIdentifier<TString>, sourceScriptRelationship: SourceScriptRelationship,
+        currentScriptRelationshipsProvider: IScriptMapperProvider): IdentifiedLoadedSource<TString> {
+
+        const contentsLocation = fs.existsSync(identifier.textRepresentation) ? ContentsLocation.PersistentStorage : ContentsLocation.DynamicMemory;
```

Comment:
Could probably make this async

Reply:
I'll add a TODO to make that change.

The method is indirectly called from the Script class constructor. I tried to make it async at certain point, but I couldn't figure out how to make the Script class constructor be async (and if I want to make the class immutable, I cannot modify the member variables after the constructor finishes).